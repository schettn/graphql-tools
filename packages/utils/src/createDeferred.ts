// TODO: Remove this after Node 22

export type Deferred<T = unknown> = PromiseWithResolvers<T>;

export function createDeferred<T>(): Deferred<T> {
  if (Promise.withResolvers) {
    return Promise.withResolvers();
  }
  let resolve: (value: T | PromiseLike<T>) => void;
  let reject: (error: unknown) => void;
  const promise = new Promise<T>((_resolve, _reject) => {
    resolve = _resolve;
    reject = _reject;
  });
  return { promise, resolve: resolve!, reject: reject! };
}
