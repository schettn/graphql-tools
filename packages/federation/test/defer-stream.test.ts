import { inspect } from 'util';
import { GraphQLSchema, parse, print } from 'graphql';
import _ from 'lodash';
import { IntrospectAndCompose, LocalGraphQLDataSource } from '@apollo/gateway';
import { buildSubgraphSchema } from '@apollo/subgraph';
import { createDefaultExecutor } from '@graphql-tools/delegate';
import { normalizedExecutor } from '@graphql-tools/executor';
import { ExecutionResult, mergeIncrementalResult } from '@graphql-tools/utils';
import { assertAsyncIterable } from '../../loaders/url/tests/test-utils';
import { getStitchedSchemaFromSupergraphSdl } from '../src/supergraph';

function mergeIncrementalResults(values: ExecutionResult[]) {
  const result: ExecutionResult = {};
  for (const value of values) {
    mergeIncrementalResult({ incrementalResult: value, executionResult: result });
  }
  return result;
}

describe('Defer/Stream', () => {
  const users = [
    { id: '1', name: 'Ada Lovelace' },
    { id: '2', name: 'Alan Turing' },
  ];
  const posts = [
    { id: '1', title: 'Hello, World!', authorId: '1' },
    { id: '2', title: 'My Story', authorId: '2' },
  ];
  function resolveWithDelay<T>(value: () => T, delay: number): Promise<T> {
    return new Promise(resolve => setTimeout(() => resolve(value()), delay));
  }
  const usersSubgraph = buildSubgraphSchema({
    typeDefs: parse(/* GraphQL */ `
      type Query {
        user(id: ID!): User
        users: [User]
        usersStream: [User]
      }

      type User @key(fields: "id") {
        id: ID!
        name: String!
      }
    `),
    resolvers: {
      User: {
        __resolveReference: (user: { id: string }) =>
          resolveWithDelay(() => users.find(u => u.id === user.id), 500),
      },
      Query: {
        users: () => resolveWithDelay(() => users, 300),
        user: (_, { id }) => resolveWithDelay(() => users.find(user => user.id === id), 500),
        usersStream: async function* usersStream() {
          for (const user of users) {
            yield user;
            await new Promise(resolve => setTimeout(resolve, 500));
          }
        },
      },
    },
  });
  const postsSubgraph = buildSubgraphSchema({
    typeDefs: parse(/* GraphQL */ `
      type Query {
        post(id: ID!): Post
        posts: [Post]
      }

      type Post @key(fields: "id") {
        id: ID!
        title: String
        author: User!
      }

      extend type User @key(fields: "id") {
        id: ID! @external
        posts: [Post]
      }
    `),
    resolvers: {
      Post: {
        __resolveReference: (post: { id: string }) =>
          resolveWithDelay(() => posts.find(p => p.id === post.id), 300),
        author: (post: { authorId: string }) =>
          resolveWithDelay(() => users.find(user => user.id === post.authorId), 300),
      },
      User: {
        __resolveReference: (user: { id: string }) =>
          resolveWithDelay(() => users.find(u => u.id === user.id), 300),
        posts: (user: { id: string }) =>
          resolveWithDelay(() => posts.filter(post => post.authorId === user.id), 300),
      },
      Query: {
        post: (_, { id }) => resolveWithDelay(() => posts.find(post => post.id === id), 300),
        posts: () => resolveWithDelay(() => posts, 300),
        postsStream: async function* postsStream() {
          for (const post of posts) {
            yield post;
            await new Promise(resolve => setTimeout(resolve, 500));
          }
        },
      },
    },
  });
  let schema: GraphQLSchema;
  let finalResult: ExecutionResult;
  beforeAll(async () => {
    const { supergraphSdl } = await new IntrospectAndCompose({
      subgraphs: [
        { name: 'users', url: 'http://localhost:4001/graphql' },
        { name: 'posts', url: 'http://localhost:4002/graphql' },
      ],
    }).initialize({
      update() {},
      async healthCheck() {},
      getDataSource({ name }) {
        if (name === 'users') return new LocalGraphQLDataSource(usersSubgraph);
        if (name === 'posts') return new LocalGraphQLDataSource(postsSubgraph);
        throw new Error(`Unknown subgraph: ${name}`);
      },
    });
    schema = getStitchedSchemaFromSupergraphSdl({
      supergraphSdl,
      onSubschemaConfig(subschemaConfig) {
        const subgraphName = subschemaConfig.name.toLowerCase();
        if (subgraphName === 'users') {
          const origExecutor = createDefaultExecutor(usersSubgraph);
          subschemaConfig.executor = async function usersExecutor(execReq) {
            const result = await origExecutor(execReq);
            if (process.env['DEBUG']) {
              console.log({
                subgraphName,
                document: print(execReq.document),
                result: inspect(result, false, Infinity),
              });
            }
            return result;
          };
        } else if (subgraphName === 'posts') {
          const origExecutor = createDefaultExecutor(postsSubgraph);
          subschemaConfig.executor = async function postsExecutor(execReq) {
            const result = await origExecutor(execReq);
            if (process.env['DEBUG']) {
              console.log({
                subgraphName,
                document: print(execReq.document),
                result: inspect(result, false, Infinity),
              });
            }
            return result;
          };
        } else {
          throw new Error(`Unknown subgraph: ${subgraphName}`);
        }
      },
    });
    finalResult = (await normalizedExecutor({
      schema,
      document: parse(/* GraphQL */ `
        query {
          users {
            id
            name
            posts {
              id
              title
              author {
                id
                name
              }
            }
          }
          posts {
            id
            title
            author {
              id
              name
              posts {
                id
                title
              }
            }
          }
        }
      `),
    })) as ExecutionResult;
  });
  it('defers the root fields', async () => {
    const result = await normalizedExecutor({
      schema,
      document: parse(/* GraphQL */ `
        query {
          ... @defer {
            users {
              id
              name
              posts {
                id
                title
                author {
                  id
                  name
                }
              }
            }
          }
          ... @defer {
            posts {
              id
              title
              author {
                id
                name
                posts {
                  id
                  title
                }
              }
            }
          }
        }
      `),
    });
    assertAsyncIterable(result);
    const values = [];
    for await (const value of result) {
      if (value.incremental?.some(v => v.errors?.length)) {
        throw new Error(`Unexpected on incremental response: ${inspect(value, false, Infinity)}`);
      }
      if (value.errors) {
        throw new Error(`Unexpected error: ${value.errors}`);
      }
      values.push(value);
    }
    expect(values).toMatchSnapshot('defer-root-fields');
    const mergedResult = mergeIncrementalResults(values);
    expect(mergedResult).toEqual(finalResult);
  });
  it('defers the nested fields', async () => {
    const result = await normalizedExecutor({
      schema,
      document: parse(/* GraphQL */ `
        query {
          users {
            id
            name
            ... @defer {
              posts {
                id
                ... @defer {
                  title
                }
                author {
                  id
                  ... @defer {
                    name
                  }
                }
              }
            }
          }
          ... @defer {
            posts {
              id
              title
              author {
                id
                ... @defer {
                  name
                }
                posts {
                  id
                  title
                }
              }
            }
          }
        }
      `),
    });
    assertAsyncIterable(result);
    const values = [];
    for await (const value of result) {
      if (value.incremental?.some(v => v.errors?.length)) {
        throw new Error(`Unexpected on incremental response: ${inspect(value, false, Infinity)}`);
      }
      if (value.errors) {
        throw new Error(`Unexpected error: ${value.errors}`);
      }
      values.push(value);
    }
    expect(values).toMatchSnapshot('defer-nested-fields');
    const mergedResult = mergeIncrementalResults(values);
    expect(mergedResult).toEqual(finalResult);
  });
  it('streams', async () => {
    const result = await normalizedExecutor({
      schema,
      document: parse(/* GraphQL */ `
        query {
          usersStream @stream {
            id
            name
            posts {
              id
              title
              author {
                id
                name
              }
            }
          }
        }
      `),
    });
    assertAsyncIterable(result);
    const values = [];
    for await (const value of result) {
      if (value.incremental?.some(v => v.errors?.length)) {
        throw new Error(`Unexpected on incremental response: ${inspect(value, false, Infinity)}`);
      }
      if (value.errors) {
        throw new Error(`Unexpected error: ${value.errors}`);
      }
      values.push(value);
    }
    expect(values).toMatchSnapshot('stream');
    const mergedResult = mergeIncrementalResults(values);
    expect(mergedResult).toEqual({
      data: {
        usersStream: [
          {
            id: '1',
            name: 'Ada Lovelace',
            posts: [
              {
                id: '1',
                title: 'Hello, World!',
                author: {
                  id: '1',
                  name: 'Ada Lovelace',
                },
              },
            ],
          },
          {
            id: '2',
            name: 'Alan Turing',
            posts: [
              {
                id: '2',
                title: 'My Story',
                author: {
                  id: '2',
                  name: 'Alan Turing',
                },
              },
            ],
          },
        ],
      },
    });
  });
});
