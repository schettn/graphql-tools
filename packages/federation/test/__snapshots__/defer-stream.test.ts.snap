// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Defer/Stream defers the nested fields: defer-nested-fields 1`] = `
[
  {
    "data": {
      "users": [
        {
          "id": "1",
          "name": "Ada Lovelace",
        },
        {
          "id": "2",
          "name": "Alan Turing",
        },
      ],
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [],
      },
      {
        "id": "1",
        "path": [
          "users",
          0,
        ],
      },
      {
        "id": "2",
        "path": [
          "users",
          1,
        ],
      },
    ],
  },
  {
    "completed": [
      {
        "id": "0",
      },
      {
        "id": "3",
      },
      {
        "id": "4",
      },
    ],
    "hasNext": true,
    "incremental": [
      {
        "data": {
          "posts": [
            {
              "author": {
                "id": "1",
                "posts": [
                  {
                    "id": "1",
                    "title": "Hello, World!",
                  },
                ],
              },
              "id": "1",
              "title": "Hello, World!",
            },
            {
              "author": {
                "id": "2",
                "posts": [
                  {
                    "id": "2",
                    "title": "My Story",
                  },
                ],
              },
              "id": "2",
              "title": "My Story",
            },
          ],
        },
        "id": "0",
      },
      {
        "data": {
          "name": "Ada Lovelace",
        },
        "id": "3",
      },
      {
        "data": {
          "name": "Alan Turing",
        },
        "id": "4",
      },
    ],
    "pending": [
      {
        "id": "3",
        "path": [
          "posts",
          0,
          "author",
        ],
      },
      {
        "id": "4",
        "path": [
          "posts",
          1,
          "author",
        ],
      },
    ],
  },
  {
    "completed": [
      {
        "id": "1",
      },
      {
        "id": "2",
      },
      {
        "id": "5",
      },
      {
        "id": "6",
      },
      {
        "id": "7",
      },
      {
        "id": "8",
      },
    ],
    "hasNext": false,
    "incremental": [
      {
        "data": {
          "posts": [
            {
              "author": {
                "id": "1",
              },
              "id": "1",
            },
          ],
        },
        "id": "1",
      },
      {
        "data": {
          "posts": [
            {
              "author": {
                "id": "2",
              },
              "id": "2",
            },
          ],
        },
        "id": "2",
      },
      {
        "data": {
          "title": "Hello, World!",
        },
        "id": "5",
      },
      {
        "data": {
          "name": "Ada Lovelace",
        },
        "id": "6",
      },
      {
        "data": {
          "title": "My Story",
        },
        "id": "7",
      },
      {
        "data": {
          "name": "Alan Turing",
        },
        "id": "8",
      },
    ],
    "pending": [
      {
        "id": "5",
        "path": [
          "users",
          0,
          "posts",
          0,
        ],
      },
      {
        "id": "6",
        "path": [
          "users",
          0,
          "posts",
          0,
          "author",
        ],
      },
      {
        "id": "7",
        "path": [
          "users",
          1,
          "posts",
          0,
        ],
      },
      {
        "id": "8",
        "path": [
          "users",
          1,
          "posts",
          0,
          "author",
        ],
      },
    ],
  },
]
`;

exports[`Defer/Stream defers the root fields: defer-root-fields 1`] = `
[
  {
    "data": {},
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [],
      },
      {
        "id": "1",
        "path": [],
      },
    ],
  },
  {
    "completed": [
      {
        "id": "0",
      },
    ],
    "hasNext": true,
    "incremental": [
      {
        "data": {
          "users": [
            {
              "id": "1",
              "name": "Ada Lovelace",
              "posts": [
                {
                  "author": {
                    "id": "1",
                    "name": "Ada Lovelace",
                  },
                  "id": "1",
                  "title": "Hello, World!",
                },
              ],
            },
            {
              "id": "2",
              "name": "Alan Turing",
              "posts": [
                {
                  "author": {
                    "id": "2",
                    "name": "Alan Turing",
                  },
                  "id": "2",
                  "title": "My Story",
                },
              ],
            },
          ],
        },
        "id": "0",
      },
    ],
  },
  {
    "completed": [
      {
        "id": "1",
      },
    ],
    "hasNext": false,
    "incremental": [
      {
        "data": {
          "posts": [
            {
              "author": {
                "id": "1",
                "name": "Ada Lovelace",
                "posts": [
                  {
                    "id": "1",
                    "title": "Hello, World!",
                  },
                ],
              },
              "id": "1",
              "title": "Hello, World!",
            },
            {
              "author": {
                "id": "2",
                "name": "Alan Turing",
                "posts": [
                  {
                    "id": "2",
                    "title": "My Story",
                  },
                ],
              },
              "id": "2",
              "title": "My Story",
            },
          ],
        },
        "id": "1",
      },
    ],
  },
]
`;

exports[`Defer/Stream streams: stream 1`] = `
[
  {
    "data": {
      "usersStream": [],
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "usersStream",
        ],
      },
    ],
  },
  {
    "completed": [
      {
        "id": "0",
      },
    ],
    "hasNext": false,
    "incremental": [
      {
        "id": "0",
        "items": [
          {
            "id": "1",
            "name": "Ada Lovelace",
            "posts": [
              {
                "author": {
                  "id": "1",
                  "name": "Ada Lovelace",
                },
                "id": "1",
                "title": "Hello, World!",
              },
            ],
          },
          {
            "id": "2",
            "name": "Alan Turing",
            "posts": [
              {
                "author": {
                  "id": "2",
                  "name": "Alan Turing",
                },
                "id": "2",
                "title": "My Story",
              },
            ],
          },
        ],
      },
    ],
  },
]
`;
